### PR TITLE
PLAN-156-late-user-joins-the-game

### PR DESCRIPTION
**Pull request by AndzejIlj**

### CODE REVIEW

File: services/api/src/messaging/players.service.ts [hunk 1]
```diff
@@ -186,6 +240,19 @@ export const playerConnected: Handler = (
...
+  sendMessage(
+    ws,
+    MessageType.UpdateGameState,
+    gameRoomService.getGameState(ws.roomId),
+  );
+
+  if (gameRoomService.getGameState(ws.roomId) === GameState.GAME_NOT_STARTED) {
+    if (gameService.getVoters(ws.roomId).length > 1) {
+      gameRoomService.setGameState(ws.roomId, GameState.GAME_IN_PROGRESS);
+      publishGameState(ws);
```
Comment (jvalatkevicius) [inline services/api/src/messaging/players.service.ts:252]: Could we move this call into a single place and it would be executed after every action taken by players?

File: services/api/src/messaging/players.service.ts [hunk 2]
```diff
@@ -232,6 +299,42 @@ export const subscribe = (ws: GameWebSocket, newPlayer: Player): void => {
 
 export const unsubscribe = (ws: GameWebSocket): void => {
   const players = getPlayers(ws.roomId);
+  try {
```
Comment (jvalatkevicius) [inline services/api/src/messaging/players.service.ts:302]: This method got too large. Its hard to follow.

File: services/api/src/messaging/players.service.ts [hunk 3]
```diff
@@ -232,6 +299,42 @@ export const subscribe = (ws: GameWebSocket, newPlayer: Player): void => {
...
+    const playerId = findPlayerByConnection(ws).id;
+
+    if (
+      gameService.getVoters(ws.roomId).length === 2 &&
```
Comment (jvalatkevicius) [inline services/api/src/messaging/players.service.ts:306]: why is this condition needed?

File: services/api/src/messaging/players.service.ts [hunk 4]
```diff
@@ -232,6 +299,42 @@ export const subscribe = (ws: GameWebSocket, newPlayer: Player): void => {
...
+      gameRoomService.getGameState(ws.roomId) !== GameState.GAME_FINISHED
```
Comment (jvalatkevicius) [inline services/api/src/messaging/players.service.ts:307]: Can we create a general algorithm that would be applied no matter how many players we have in the game or what the game state is.

File: services/api/src/messaging/players.service.ts [hunk 5]
```diff
@@ -232,6 +299,42 @@ export const subscribe = (ws: GameWebSocket, newPlayer: Player): void => {
...
+          logger.error(err?.message);
+          logger.info('Publishing: player disconnected the game.');
+          gameRoomService.setGameState(ws.roomId, GameState.GAME_FINISHED);
+          const allPlayers = Array.from(players.values());
+          publish(ws.roomId, {
+            type: MessageType.PlayerDisconnected,
+            payload: allPlayers,
+          });
+          publishGameState(ws);
+          publishFinalBoard(ws);
```
Comment (jvalatkevicius) [inline services/api/src/messaging/players.service.ts:329]: This is repeated all over the place. Can we find a better way to determine when final board should be pushed apart from placing it directly in the code?

File: services/api/src/messaging/players.service.ts [hunk 6]
```diff
@@ -146,6 +196,10 @@ const createNewPlayer = (params: {
     newPlayer.status = PlayerStatus.FigurePlaced;
   } else if (gameService.playerHasSkipped(params.roomId, params.playerId)) {
     newPlayer.status = PlayerStatus.MoveSkipped;
+  } else if (
```
Comment (jvalatkevicius) [inline services/api/src/messaging/players.service.ts:199]: theres no `if` condition

File: services/api/src/messaging/players.service.ts [hunk 7]
```diff
@@ -80,6 +120,12 @@ const setDefaultStatusForPlayers = (ws: GameWebSocket): void => {
 
 export const resetGame = (ws: GameWebSocket): void => {
   gameService.clearBoard(ws.roomId);
+  if (gameService.getVoters(ws.roomId).length < 2) {
```
Comment (jvalatkevicius) [inline services/api/src/messaging/players.service.ts:123]: why is this condition needed?

File: services/api/src/messaging/players.service.ts [hunk 8]
```diff
@@ -54,7 +82,18 @@ const publishFinalBoard = (ws: GameWebSocket): void => {
...
 
+const publishGameState = (ws: GameWebSocket): void => {
+  publish(ws.roomId, {
+    type: MessageType.UpdateGameState,
+    payload: gameRoomService.getGameState(ws.roomId),
+  });
+};
+
 export const figureMoved: Handler = (ws, payload: PlaceFigureMessage): void => {
+  if (gameRoomService.getGameState(ws.roomId) !== GameState.GAME_IN_PROGRESS) {
```
Comment (jvalatkevicius) [inline services/api/src/messaging/players.service.ts:93]: this seems like some sort of workaround. Why is this needed?

File: services/api/src/messaging/players.service.ts [hunk 9]
```diff
@@ -36,6 +37,33 @@ export const findPlayerById = (
...
+): boolean => {
+  const player = findPlayerById(roomId, playerId);
+  const voters = gameService.getVoters(roomId);
+  const votersWhoMadeMoveCount = gameService.getVoterWhoMadeActionCount(roomId);
+  const playerCount = gameRoomService.getPlayers(roomId).size;
+
+  const isOneVoteMissing = playerCount === votersWhoMadeMoveCount + 1;
+  const hasPlayerMoved = player[1].status !== PlayerStatus.ActionNotTaken;
+
+  if (isOneVoteMissing && voters.length > 1 && !hasPlayerMoved) {
```
Comment (jvalatkevicius) [inline services/api/src/messaging/players.service.ts:61]: why `isOneVoteMissing` not enough?

File: services/api/src/messaging/players.service.ts [hunk 10]
```diff
@@ -36,6 +37,33 @@ export const findPlayerById = (
...
+    throw new Error('player with this ws connection not found');
+  }
+  return player;
+};
+
+export const checkIfLastToVote = (
+  roomId: string,
+  playerId: string,
+): boolean => {
+  const player = findPlayerById(roomId, playerId);
```
Comment (jvalatkevicius) [inline services/api/src/messaging/players.service.ts:53]: What is the logic behind puting methods into services?
Why is this method inside the players service, while `getPlayers`  and `getVoterWhoMadeActionCount` are in gameService and gameRoomService?

**Review by jvalatkevicius — CHANGES_REQUESTED** [review]

File: services/api/src/messaging/players.service.ts (hunk 3 above)
Comment (AndzejIlj) [reply services/api/src/messaging/players.service.ts:306]: Because, when there is only to people in the room and the game is not finished, when one of them logs out, in my opinion the game state should become game not started (we don't allow player to make votes, when he's the only one in the room).

File: services/api/src/messaging/players.service.ts (hunk 6 above)
Comment (AndzejIlj) [reply services/api/src/messaging/players.service.ts:199]: There is, its in the next line and wrapped with ( ). It probably got formatted this way, because the line got too large

File: services/api/src/messaging/players.service.ts (hunk 7 above)
Comment (AndzejIlj) [reply services/api/src/messaging/players.service.ts:123]: Because, without this condition, if one player joins the room and presses restart game button, the game state would become GameInProgress and he would be able to place figures on the board (at least in FE)

File: services/api/src/messaging/players.service.ts (hunk 8 above)
Comment (AndzejIlj) [reply services/api/src/messaging/players.service.ts:93]: Its just for safety reasons. Even if the game state would be GameNotStarted and user could not place figure on the board, we could send a message with a bot, and probably crash the game.

File: services/api/src/messaging/players.service.ts (hunk 9 above)
Comment (AndzejIlj) [reply services/api/src/messaging/players.service.ts:61]: I've got rid of `voters.length > 1` condition, but `hasPlayerMoved` is needed, because, we need to know, if the particular player has not moved. It also reflects in method name, which is `checkIfLastToVote`

File: services/api/src/messaging/players.service.ts (hunk 8 above)
Comment (AndzejIlj) [reply services/api/src/messaging/players.service.ts:93]: Its for a safety reasons. Without this check, even if game state is `GameNotStarted` and user cannot place figure on the board, it would be possible to send a message for example with a bot

File: services/api/src/messaging/players.service.ts (hunk 1 above)
Comment (AndzejIlj) [reply services/api/src/messaging/players.service.ts:252]: Why should we send so many messages? For example if user makes two turns(for example changes his mind), we would publish game that has not changed state twice. I think we should leave game state publishing, to when its really needed or game state changes
